### PR TITLE
replace EventFd with Trigger abstraction

### LINKED ISSUE #7: Add abstraction for notifications instead of EventFd
#1 uses an EventFd for notifying the driver about events. We should replace this rather specific mechanism with some abstraction.

### PR DESCRIPTION
**Pull request by lauralt**
Fixes #7.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Changelog 
+## [Unreleased]
```
Comment (andreeaflorescu) [inline CHANGELOG.md:2]: nit: add a whitespace.

File: src/trigger.rs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (andreeaflorescu) [inline src/trigger.rs:1]: I find it a bit weird to have a module with just one trait. How about leaving it to src/lib.rs for now? In the future we can move it to a specific module (for example one with `events` if we find such a module useful). It is subjective, but I think the overhead of using it from a separate module defeats the separation purpose.

File: src/trigger.rs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+//! Provides an interface for an object that can generate an event.
+
+use std::result::Result;
+
+/// Abstraction for a simple, push-button like interrupt mechanism.
```
Comment (andreeaflorescu) [inline src/trigger.rs:8]: Besides what it is, I think it would be important to add some details about the "why". Why did you add this trait? What are you trying to abstract? When are developers of this crate (or users of this interface) suppose to use it?

File: src/serial.rs [hunk 4]
```diff
@@ -144,16 +153,41 @@ impl<W: Write> Serial<W> {
...
-    /// # use std::io::sink;
+    /// # use std::io::{sink, Error, Result};
+    /// # use std::ops::Deref;
+    /// # use vm_superio::trigger::Trigger;
     /// # use vm_superio::Serial;
     /// # use vmm_sys_util::eventfd::EventFd;
-    /// let intr_evt = EventFd::new(libc::EFD_NONBLOCK).unwrap();
-    /// let serial = Serial::new(intr_evt.try_clone().unwrap(), Vec::new());
     ///
+    /// struct TestEventFd(EventFd);
```
Comment (andreeaflorescu) [inline src/serial.rs:162]: I wouldn't name it `Test..` in this context because this shows up in the public documentation. How about `EventFdTrigger`?

File: src/serial.rs [hunk 5]
```diff
@@ -191,8 +225,8 @@ impl<W: Write> Serial<W> {
         (self.modem_control & MCR_LOOP_BIT) != 0
     }
 
-    fn trigger_interrupt(&mut self) -> Result<()> {
-        self.interrupt_evt.write(1)
+    fn trigger_interrupt(&mut self) -> std::result::Result<(), T::E> {
```
Comment (andreeaflorescu) [inline src/serial.rs:228]: Since we are using this std::result::Result quite a lot, how about we define a `Result` wrapper type?

File: src/serial.rs [hunk 6]
```diff
@@ -254,15 +288,38 @@ impl<W: Write> Serial<W> {
...
     /// ```rust
+    /// # use std::io::{Error, Result};
+    /// # use std::ops::Deref;
+    /// # use vm_superio::trigger::Trigger;
     /// # use vm_superio::Serial;
     /// # use vmm_sys_util::eventfd::EventFd;
-    /// let intr_evt = EventFd::new(libc::EFD_NONBLOCK).unwrap();
-    /// let mut serial = Serial::new(intr_evt, Vec::new());
     ///
+    /// struct TestEventFd(EventFd);
```
Comment (andreeaflorescu) [inline src/serial.rs:297]: Duplicating this for all public functions is a bit of a bummer. What I saw other projects do instead is to provide just one example at the module level with all public functions, and then just point to that example in the function documentation.

Can we also do this here?

File: src/serial.rs [hunk 7]
```diff
@@ -383,18 +465,43 @@ impl<W: Write> Serial<W> {
...
     /// serial.enqueue_raw_bytes(&[b'a', b'b', b'c']).unwrap();
     /// ```
-    pub fn enqueue_raw_bytes(&mut self, input: &[u8]) -> Result<()> {
+    pub fn enqueue_raw_bytes(&mut self, input: &[u8]) -> std::result::Result<(), Error<T::E>> {
         if !self.is_in_loop_mode() {
             self.in_buffer.extend(input);
             self.set_lsr_rda_bit();
-            self.received_data_interrupt()?;
+            self.received_data_interrupt()
+                .map_err(Error::TriggerInterrupt)?;
```
Comment (andreeaflorescu) [inline src/serial.rs:504]: How about we implement `From` for `Error` so that we do not need to call `map_err`?

File: src/serial.rs (hunk 6 above)
Comment (lauralt) [reply src/serial.rs:297]: Good suggestion, I'll try to see how it looks.

File: src/serial.rs [hunk 8]
```diff
@@ -124,18 +124,27 @@ pub struct Serial<W: Write> {
...
+    interrupt_evt: T,
     out: W,
 }
 
-impl<W: Write> Serial<W> {
+/// Errors encountered while handling serial console operations.
+#[derive(Debug)]
+pub enum Error<E> {
+    /// Failed to trigger interrupt.
+    TriggerInterrupt(E),
```
Comment (alexandruag) [inline src/serial.rs:135]: How about we rename this variant to simply `Trigger`?

File: src/i8042.rs [hunk 9]
```diff
@@ -24,15 +25,15 @@ const CMD_RESET_CPU: u8 = 0xFE;
 
 /// Errors encountered while handling i8042 operations.
 #[derive(Debug)]
-pub enum Error {
+pub enum Error<E> {
     /// Failed to trigger interrupt.
-    TriggerInterrupt(io::Error),
+    TriggerInterrupt(E),
```
Comment (alexandruag) [inline src/i8042.rs:30]: While this is the only error which can occur here, how about we don't define an `Error` enum, and simply return something like `Result<something, T::E>` from the methods that call `trigger` below?

File: src/serial.rs (hunk 7 above)
Comment (lauralt) [reply src/serial.rs:504]: This turned out to be a bit tricky, since `Error` enum has the generic `E` type, which can be `io::Error` too (the same as the other variant's parameter). Also, the fact that we are returning here `Trigger::E` which is less generic than the other `E` seemed to make things weird. As for the i8042, I removed, at Alex's suggestion, the Error enum for now.

If you think it makes sense the effort to come up with a `From` impl here or some other errors improvement, we could log an issue, what do you say?

File: src/trigger.rs (hunk 2 above)
Comment (lauralt) [reply src/trigger.rs:1]: No strong preference, I moved it to `lib.rs`.

File: src/trigger.rs (hunk 3 above)
Comment (lauralt) [reply src/trigger.rs:8]: I added more details here, hope it's better.

**Review by lauralt — COMMENTED** [review]
@andreeaflorescu @alexandruag thanks for the review, I addressed your comments and also improved a bit the i8042 unit tests in a separate commit (the last one).

**Review by alexandruag — APPROVED** [review]
Looks good to me.

File: src/serial.rs [hunk 10]
```diff
@@ -124,36 +172,40 @@ pub struct Serial<W: Write> {
...
     // Used for notifying the driver about some in/out events.
-    interrupt_evt: EventFd,
+    interrupt_evt: T,
     out: W,
 }
 
-impl<W: Write> Serial<W> {
+/// Errors encountered while handling serial console operations.
+#[derive(Debug)]
+pub enum Error<E> {
```
Comment (aghecenco) [inline src/serial.rs:181]: I think this could benefit from being more explicitly parameterized as 
```rust
pub enum Error<T: Trigger> {
    Trigger(T::E)
```

File: src/serial.rs [hunk 11]
```diff
@@ -277,12 +322,12 @@ impl<W: Write> Serial<W> {
                     if self.in_buffer.len() < LOOP_SIZE {
                         self.in_buffer.push_back(value);
                         self.set_lsr_rda_bit();
-                        self.received_data_interrupt()?;
+                        self.received_data_interrupt().map_err(Error::Trigger)?;
```
Comment (aghecenco) [inline src/serial.rs:325]: I agree with @andreeaflorescu that these are a pain but they're unavoidable. The `From` impl would generate a conflict on the `io::Error` branch (it's basically `impl<T> From<T> for T` and rustc doesn't approve). So let's leave it this way

File: src/serial.rs (hunk 10 above)
Comment (lauralt) [reply src/serial.rs:181]: Looks like with this approach there's a problem when `T` doesn't implement `Debug` (which is the case for `EventFd`), i.e. `unwrap()` is not allowed for `std::result::Result<(), Error<T>>` (unless T implements `Debug`).
Let's keep it like it is (at least for now), if there are no other major concerns.

**Review by aghecenco — APPROVED** [review]
